Prohibit putting null values into Sessions (#8114)

MouslihAbdelhakim opened the pull request:
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [ ] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Fixes

Fixes #8114 

## Purpose

This PR prohibit users from putting null values into a Session, which will later cause a NPE upon decoding the session.

richdougherty commented on framework/src/play/src/main/scala/play/api/mvc/Session.scala:
```diff
@@ -17,6 +21,12 @@ import scala.collection.JavaConverters._
+  /**
+   * Null values in session cause NPE when decoding
+   */
+  require(!containsNullValue(data), illegalCookieValue)
```
A shorter way:

```scala
require(data.forAll(kv => kv._2 != null), "The map contained a null cookie value")
```

marcospereira replied:
@richdougherty it would be useful to have the key here, just as you suggested for the Java version. So it could be:

```scala
val nullValue = data.find(_._2 == null)
require(nullValue.isEmpty, s"The map contained a null cookie value: ${nullValue.get}")
```

richdougherty commented on framework/src/play/src/main/scala/play/api/mvc/Session.scala:
```diff
@@ -40,7 +50,7 @@ case class Session(data: Map[String, String] = Map.empty[String, String]) {
   def +(kv: (String, String)): Session = {
-    require(kv._2 != null, "Cookie values cannot be null")
+    require(kv._2 != null, illegalCookieValue)
     copy(data + kv)
```
The call to `copy` creates a new `Session` object. In the `Session` constructor there's logic to null-check all the values again. To be more efficient we want to change the `Session` class so it has a private constructor and assumes that its arguments are already valid. However this probably means changing the class so it's a normal class, not a `case class`.

Here's a sketch of how this could work...

```scala
object Session {
  // Need apply method to be backwards compatible with case class
  def apply(data: Map[String, String] = Map.empty[String, String]): Session = {
    // validate data
    new Session(data) // assumes already valid
  }
  // Needed to be backwards compatible with case class; may want to deprecate this method
  def unapply(session: Session): Option[Map[String, String]] = Some(session.data)
}

class Session private[play] (val data: Map[String, String]) {
  // assume data is valid

  def +(...): Session = {
    // validate new value
    new Session(data + newValue) // assumes valid
  }

  ...
}
```
  ...
}

MouslihAbdelhakim replied:
making the constructor private will break previous code similar to

https://github.com/playframework/playframework/blob/a0228d077f6b318a691f5d3b505f8e8d6214d558/framework/src/play/src/main/scala/play/api/mvc/Session.scala#L77

https://github.com/playframework/playframework/blob/a0228d077f6b318a691f5d3b505f8e8d6214d558/framework/src/play/src/main/scala/play/api/mvc/Session.scala#L87

richdougherty commented on framework/src/play/src/main/scala/play/api/mvc/Session.scala:
```diff
@@ -96,12 +106,19 @@ class DefaultSessionCookieBaker @Inject() (
+    this(
+      SessionConfiguration(),
+      SecretConfiguration(),
+      new CookieSignerProvider(SecretConfiguration()).get)
```
(I'm assuming these are just formatting changes.)

richdougherty commented on framework/src/play/src/main/scala/play/api/mvc/Session.scala:
```diff
@@ -158,4 +186,13 @@ object Session extends CookieBaker[Session] with FallbackCookieDataCodec {
   @deprecated("Inject play.api.mvc.SessionCookieBaker instead", "2.6.0")
   override def serialize(session: Session): Map[String, String] = session.data
+
+  private[Session] def containsNullValue(data: Map[String, String]): Boolean =
```
I don't think it's worth separating out this function logic because it's so simple to use `data.forAll(kv =>kv._2 != null)`.

richdougherty commented on framework/src/play/src/main/scala/play/api/mvc/Session.scala:
```diff
@@ -158,4 +186,13 @@ object Session extends CookieBaker[Session] with FallbackCookieDataCodec {
+      }
+      .getOrElse(false)
+
+  private[Session] val illegalCookieValue = "Cookie values cannot be null"
```
I don't think it's worth caching this string value.

MouslihAbdelhakim commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2097,6 +2099,7 @@ public String remove(Object key) {
         @Override
         public String put(String key, String value) {
             isDirty = true;
+            if (value == null) throw illegalCookieValueException;
```
(a) i think ```IllegalArgumentException``` indicated the null is an inappropriate argument, which is more suited inthis situation, it will also allow us to stay consistent with the ```+``` method from the Scala API. 

If you prefer throwing ```NullPointerException```, should i change the ```+``` method from the Scala API to throw a NPE instead of IllegalArgumentException  ?

(b) I think including the key in the exception message is a great idea.

MouslihAbdelhakim commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2118,6 +2122,15 @@ public void clear() {
+        /**
+         * Null values in session cause NPE when decoding
+         */
+        private static IllegalArgumentException illegalCookieValueException = new IllegalArgumentException("Cookie values cannot be null");
```
I was thinking more about reuse, this way the next person to work on this needs to make a change in only one place instead of several.

marcospereira replied:
So, by using `put` method as suggested in other commends should also make it easier to maintain this later, right? Also, we won't be able to have a more informative exception (with the key name) if we are always reusing the same exception instance.

marcospereira commented on framework/src/play/src/test/scala/play/api/mvc/SessionSpec.scala:
```diff
@@ -0,0 +1,21 @@
+
+class SessionSpec extends Specification {
+
+  "Session" should {
```
Add successful scenarios too.

MouslihAbdelhakim commented:
@richdougherty, @marcospereira Thank you so much for your feedback, i have implemented the code improvements you suggested.

I have implemented the same checks for the Flash interface, and i have improved the consistency of the java api's behavior related to changing the ```isDirty``` member value and the ```putAll``` method.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2118,6 +2128,15 @@ public void clear() {
 
+        private String illegalCookieValue(String key) {
+            String keyString;
+            if (key == null)
```
A comment regarding code style: the most common style in Play's codebase is:

```java
if (key == null) {
    keyString = null;
} else {
    keyString = key;
}
```
Also valid for other changes in this PR.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2118,6 +2128,15 @@ public void clear() {
+                keyString = "null";
+            else
+                keyString =  key.toString();
+            return "Cookie value for " + keyString + " must be non-null.";
```
The error message could be something like:

```java
return "Session value for key " + keyString +  " must be a non-null value. If you want to remove the value from session, use session.remove(key)";
```

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2169,6 +2197,15 @@ public void clear() {
+                keyString = "null";
+            else
+                keyString =  key.toString();
+            return "Cookie value for " + keyString + " must be non-null.";
```
Almost the same as the `Session`:

```java
return "Flash value for key " + keyString +  " must be a non-null value. If you want to remove the value from flash, use flash.remove(key)";

marcospereira commented on framework/src/play/src/main/scala/play/api/mvc/Flash.scala:
```diff
@@ -18,6 +18,16 @@ import scala.collection.JavaConverters._
+   * Null values in flash cause NPE when decoding
+   */
+  private def requireNonNullValues(data: Map[String, String]): Unit = {
+    def safeToString(key: String): String = if (key == null) "null" else key.toString
```
No need to call `toString` in a `String` object (`key`).

marcospereira commented on framework/src/play/src/test/java/play/mvc/FlashTest.java:
```diff
@@ -0,0 +1,119 @@
+        HashMap<String,String> data = new HashMap<String, String>();
+        data.put("key", "value");
+        Http.Flash flash = new Http.Flash(data);
+        boolean notEmptyAndContainsKey =  (! flash.isEmpty()) && (flash.get("key") == "value");
```
Split these checks to make the tests clearer:

```java
assertEquals(false, flash.isEmpty());
assertEquals(true, flash.containsKey("key"));
assertEquals("value", flash.get("key"));
```
The same applies to other tests.

PromanSEW replied:
@marcospereira 
`assertEquals(true, ...)` and `assertEquals(false, ...)` must be `assertTrue(...)` and `assertFalse(...)`

marcospereira commented on framework/src/play/src/test/java/play/mvc/FlashTest.java:
```diff
@@ -0,0 +1,119 @@
+    public ExpectedException exception = ExpectedException.none();
+
+    @Test
+    public void testInitializeFlashWithNullValue() {
```
I would prefer to have more descriptive test names:

```java
public void shouldThrowExceptionWhenInitializingFlashWithNullValues() {
    ...
}
```

And for the test below:

```java
public void shouldThrowExceptionWhenPuttingANullValue() {
    ...
}
```

marcospereira commented on framework/src/play/src/test/scala/play/api/mvc/FlashSpec.scala:
```diff
@@ -0,0 +1,42 @@
+      flash.isEmpty mustEqual true
+    }
+
+    "throws an illegalArgumentException when initialized with data containing a null value" in {
```
👍

MouslihAbdelhakim commented:
@marcospereira Thank you so much for the swift reply, the suggested improvements were implemented.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2118,6 +2127,12 @@ public void clear() {
         }
 
+        private String illegalCookieValue(String key) {
+            if (key == null) {
```
Since we are just concatenating `key`, this check is not needed.

marcospereira replied:
So, maybe, after that, we won't even need this method?

MouslihAbdelhakim commented:
@marcospereira you are right, that method was not needed, so i removed it from both the Scala and Java api.

richdougherty commented on framework/src/play/src/test/java/play/mvc/FlashTest.java:
```diff
@@ -0,0 +1,118 @@
+    @Test
+    public void initializeFlashWithNullKeyAndValue(){
+        HashMap<String,String> data = new HashMap<String, String>();
+        data.put(null, "value");
```
So should we allow null keys?

MouslihAbdelhakim replied:
They don't cause NPE when decoding in a play project.

@marcospereira  what do you think ?

marcospereira replied:
@richdougherty and @MouslihAbdelhakim I can't see a use case for null keys here.

richdougherty commented on framework/src/play/src/test/java/play/mvc/FlashTest.java:
```diff
@@ -0,0 +1,118 @@
+        data.put(null, "value");
+        Http.Flash flash = new Http.Flash(data);
+        assertFalse(flash.isEmpty());
+        assertEquals("value",flash.get(null));
```
Can you check the `dirty` flag for all the tests?

MouslihAbdelhakim replied:
I think it's better for tests to have a single purpose, i will see if i can add more tests to cover all the situation isDirty flag.

richdougherty commented on framework/src/play/src/test/java/play/mvc/FlashTest.java:
```diff
@@ -0,0 +1,118 @@
+        } catch (IllegalArgumentException e) { }
+        assertFalse(flash.containsKey("key"));
+    }
+}
```
Nice tests 👍

MouslihAbdelhakim replied:
Thank you.

richdougherty commented on framework/src/play/src/test/java/play/mvc/SessionTest.java:
```diff
@@ -0,0 +1,117 @@
+        } catch (Exception e ) {}
+        assertFalse(session.containsKey("key"));
+    }
+}
```
Same comments as for flash test.

richdougherty commented on framework/src/play/src/test/scala/play/api/mvc/FlashCookieSpec.scala:
```diff
@@ -103,11 +103,6 @@ class FlashCookieSpec extends org.specs2.mutable.Specification {
-    "put disallows null values" in {
-      val c = Flash(Map("foo" -> "bar"))
-      c + (("x", null)) must throwA(new IllegalArgumentException("requirement failed: Cookie values cannot be null"))
-    }
```
This test is now part of the new `FlashSpec`, right?

MouslihAbdelhakim replied:
Yes.

MouslihAbdelhakim commented on framework/src/play/src/main/scala/play/api/mvc/Session.scala:
```diff
@@ -18,6 +18,18 @@ import scala.collection.JavaConverters._
+      s"Session value for key ${pairWithNullValue.get._2}  must be a non-null value. If you want to remove the value from session, use session - key"
+    )
+  }
+  requireNonNullValues(data)
```
Do you prefer we list all the keys that have nulls ? for example:

Session value for key1, key2 and key3 contain a null value.

richdougherty replied:
I think the error message you've already done is very nice and detailed. What I was suggesting was that you could reduce the amount of checking and error code a bit if you want at the cost of having a less specific error message. But what you've done is great so feel free to leave it if you like. :)

MouslihAbdelhakim commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2076,10 +2076,12 @@ public void discardCookie(String name, String path, String domain, boolean secur
     public static class Session extends HashMap<String,String>{
 
-        public boolean isDirty = false;
+        public boolean isDirty;
```
Yes, it should be modified if a call to put, puttAll or remove.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2127,10 +2139,13 @@ public void clear() {
      */
     public static class Flash extends HashMap<String,String>{
```
I thought the same about having a common superclass.

Anyway, we can break binary compatibility at the master branch and backport the changes to 2.6.x in a binary compatible way.

richdougherty replied:
OK @marcospereira that sounds good to me. @MouslihAbdelhakim do you see what we mean here? There's some common functionality in the `Flash` and `Session` classes that can be in a common superclass. Unfortunately we can't introduce this class into our `2.6.x` branch since 2.6 is the stable branch and we don't want people's code to break when we release a new version. What we can do is make bigger changes on the `master` branch and then make a more minimal change on the `2.6.x` branch. That means that you'd make the bigger `master` change in this PR and then - after that has been merged - do a safer "backport" PR to the `2.6.x` branch.

MouslihAbdelhakim replied:
@richdougherty, @marcospereira, thank you for the feedback, i have made and commited the necessary changes and this PR points to the playframework's master branch.

>> more minimal change on the 2.6.x branch. That means that you'd make the bigger master change in this PR and then - after that has been merged - do a safer "backport" PR to the 2.6.x branch.

this means that for the ```2.6.x``` branch we fix the NPE problems but without making a common super class for the Session and Flash classes?

marcospereira replied:
> this means that for the 2.6.x branch we fix the NPE problems but without making a common super class for the Session and Flash classes?

Yes. A superclass would break binary compatibility in 2.6.x which is something we want to avoid.

PromanSEW commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2070,20 +2070,20 @@ public void discardCookie(String name, String path, String domain, boolean secur
+     * HTTP cookie storage.
      */
-    public static class Session extends HashMap<String,String>{
+    static abstract class HttpCookieStorage extends HashMap<String,String>{
```
It is breaking change of API

MouslihAbdelhakim replied:
@PromanSEW breaking binary compatibility or something else i missed?

PromanSEW replied:
What is wrong with `Session` name?

marcospereira replied:
Maybe `AbstractCookieDataStorage`? Also, this class can be `private` and make `Session` and `Flash` `final` classes. @richdougherty what do you think?

MouslihAbdelhakim replied:
@marcospereira making AbstractCookieDataStorage private results in some compilation errors (below), while making it protected does not cause any compilation errors.
```
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/java/play/mvc/Http.java:2169: private class AbstractHttpCookieStorage escapes its defining scope as part of type play.mvc.Http.AbstractHttpCookieStorage
[error]     public static final class Flash extends AbstractHttpCookieStorage {
[error]                                             ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/api/mvc/Flash.scala:127: value asScala is not a member of play.mvc.Http.Flash
[error]   def fromJavaFlash(javaFlash: play.mvc.Http.Flash): Flash = new Flash(javaFlash.asScala.toMap)
[error]                                                                                  ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/java/play/mvc/Http.java:2148: private class AbstractHttpCookieStorage escapes its defining scope as part of type play.mvc.Http.AbstractHttpCookieStorage
[error]     public static final class Session extends AbstractHttpCookieStorage {
[error]                                               ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/api/mvc/Session.scala:132: value asScala is not a member of play.mvc.Http.Session
[error]   def fromJavaSession(javaSession: play.mvc.Http.Session): Session = new Session(javaSession.asScala.toMap)
[error]                                                                                              ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:139: value isDirty is not a member of play.mvc.Http.Session
[error]     if (javaContext.session.isDirty && javaContext.flash.isDirty) {
[error]                             ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:139: value isDirty is not a member of play.mvc.Http.Flash
[error]     if (javaContext.session.isDirty && javaContext.flash.isDirty) {
[error]                                                          ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:140: value asScala is not a member of play.mvc.Http.Session
[error]       wResult.withSession(Session(javaContext.session.asScala.toMap)).flashing(Flash(javaContext.flash.asScala.toMap))
[error]                                                       ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:140: value asScala is not a member of play.mvc.Http.Flash
[error]       wResult.withSession(Session(javaContext.session.asScala.toMap)).flashing(Flash(javaContext.flash.asScala.toMap))
[error]                                                                                                        ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:142: value isDirty is not a member of play.mvc.Http.Session
[error]       if (javaContext.session.isDirty) {
[error]                               ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:143: value asScala is not a member of play.mvc.Http.Session
[error]         wResult.withSession(Session(javaContext.session.asScala.toMap))
[error]                                                         ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:145: value isDirty is not a member of play.mvc.Http.Flash
[error]         if (javaContext.flash.isDirty) {
[error]                               ^
[error] /Users/gandalf/IdeaProjects/playframework/framework/src/play/src/main/scala/play/core/j/JavaHelpers.scala:146: value asScala is not a member of play.mvc.Http.Flash
[error]           wResult.flashing(Flash(javaContext.flash.asScala.toMap))
```

PromanSEW commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2092,83 +2092,65 @@ public String remove(Object key) {
+     * <p>
+     * Session data are encoded into an HTTP cookie, and can only contain simple <code>String</code> values.
+     */
+    public static class Session extends HttpCookieStorage {
```
I see why you renamed this. OK

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2092,83 +2092,65 @@ public String remove(Object key) {
         public String put(String key, String value) {
+            if (key == null || value == null) throw
+                    new IllegalArgumentException(
+                            "the key: " + key +  " and the value: " + value +" must be non-null."
```
Don't think it is needed since it is null. Maybe handle this separately so that we can have super clear exception messages:

```java
if (key == null) {
    throw new IllegalArgumentException("key must not be null when adding session data");
}

if (value == null) {
    throw new IllegalArgumentException("value must not be null when adding session data");
}
```
By the way, I think subclasses can have something like `reportError` so that we can differentiate between session and flash. Something like:

```java
if (key == null || value == null) reportError(key, value);
```

And inside `Session`:

```java
public void reportError(String key, String value) {
  throw new IllegalArgumentException(
    String.format("Null key or value when adding session data: (%s, %s)", key, value)
  );
}
```
What do you think?

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2092,83 +2092,95 @@ public String remove(Object key) {
          */
         @Override
         public String put(String key, String value) {
+            if (key == null) throw
```
I think you can extract this to a private method and then reuse it in `putAll` too.

marcospereira commented on framework/src/play/src/main/java/play/mvc/Http.java:
```diff
@@ -2092,83 +2092,95 @@ public String remove(Object key) {
             super.clear();
         }
 
+        /**
```
Remove the empty comments.

marcospereira commented:
Hi @MouslihAbdelhakim, we are planning to do a release next week and it would be good to have this as part of it.

Could you please resolve the conflicts so that we can move this PR forward?

Thank you!

MouslihAbdelhakim commented:
@marcospereira done.

mkurz commented:
@MouslihAbdelhakim Can you please once more  resolve the conflict? Thank you very much!

MouslihAbdelhakim commented:
@mkurz i'll gladly do it.

One question though, Can this PR be merged without @richdougherty approving his change requests ?

mkurz commented:
@MouslihAbdelhakim Yes this is possible. Only one positiv review is required to merge a pull request. And since @marcospereira already approved this pull request we can merge it without rich's approval.

MouslihAbdelhakim commented:
@mkurz done (:

mkurz commented:
@MouslihAbdelhakim Something went wrong here. Can you rebase again using `git rebase`?

mkurz commented:
@MouslihAbdelhakim sorry I just saw you wrote me a message via gitter. I just replied to it.

mkurz commented:
FYI: This PR has been abandoned. Looks like the author has no time to work on this and (unfortunately) also deleted the repo.

dwijnand commented:
Note that when you open a pull request the target repo has an internal git reference for that branch.  So even if the repo was deleted you can still access the branch.  The easiest way is by using https://hub.github.com/ and `hub checkout https://github.com/playframework/playframework/pull/8141`.
